### PR TITLE
feat(generate): support `HEAD --changelog` as `[Unreleased]` target

### PR DESCRIPTION
**Pull request by ThomasK33**
Hey jdx, really cool project, thanks for kicking it off. What do you think of allowing someone to generate a changelog entry for all unreleased entries if they generate changelogs off HEAD? I took the liberty of cooking up a PoC in this PR. Below are the agent's PR description and plan used to cook this up.

---

## Summary

Make `communique generate HEAD --changelog` an intentional way to produce/refresh the `[Unreleased]` section of `CHANGELOG.md`, instead of writing a literal `## [HEAD]` entry.

Today the command already collects the right git range (latest tag → `HEAD`), but it labels the result as a release named `HEAD` and inserts `## [HEAD]` into the changelog. This PR keeps the existing range mechanics and instead routes `HEAD` to "Unreleased" semantics end-to-end.

## Motivation

When iterating on unreleased work, it's useful to regenerate the `[Unreleased]` block from current `main` without first creating a tag or hand-editing the changelog. A literal `## [HEAD]` heading is never what a user wants — but the rest of the pipeline is already a great fit, so the change is mostly a matter of relabeling and a deterministic in-place changelog edit.

## Behavior

- `communique generate HEAD --changelog`
  - Replaces the body of the existing `## [Unreleased]` (or `## Unreleased`) section in place.
  - Preserves the file preamble, the exact unreleased header text as found, and all released sections that follow.
  - If `CHANGELOG.md` doesn't exist, creates it with `# Changelog`, `## [Unreleased]`, and the generated body.
  - If an existing `CHANGELOG.md` has no Unreleased header, fails with an actionable `miette` diagnostic instead of guessing where to insert content.
- Prompts and user-facing labels in `HEAD` mode read as "unreleased changes since {prev_tag}" rather than `**HEAD**`. Existing `[Unreleased]` content is passed in as draft material to reconcile, and the model is instructed not to emit a nested `## [Unreleased]` heading.
- `communique generate HEAD --github-release` now fails fast with a clear message — `HEAD` is not a release tag and shouldn't be searched for/updated on GitHub.
- Tagged-release behavior (`vX.Y.Z --changelog`, `vX.Y.Z --github-release`, prompt phrasing, title normalization) is unchanged.

The literal `unreleased` CLI argument is intentionally **not** added in this PR; `HEAD` is the only new entry point.

## Implementation notes

- New helpers on `Context`: `is_unreleased_head()` and `display_tag()` to keep the literal `HEAD` from leaking into user-facing output while preserving `ctx.tag` for git ref usage.
- Deterministic, line-based `[Unreleased]` section replacement — no second LLM call to merge the section. Handles both `## [Unreleased]` and `## Unreleased` headers.
- `--github-release` guard runs early, before any GitHub API call.
- `debug_assert!`s used only for internal invariants in the HEAD-only branch; user-facing failure modes use `miette` errors.

## Files touched

| File | Net | Purpose |
| --- | --- | --- |
| `src/generate.rs` | +463 | Helpers, HEAD branch in `update_changelog`, deterministic section replacer, `--github-release` guard, focused unit tests |
| `src/prompt.rs` | +88 | `is_unreleased_head` prompt context + unreleased phrasing |
| `tests/integration.rs` | +102 | New mocked end-to-end test `test_generate_head_changelog_updates_unreleased` |

## Testing

New tests:

- `test_update_changelog_head_updates_unreleased_in_place`
- `test_update_changelog_head_creates_missing_file`
- `test_update_changelog_head_errors_without_unreleased_header`
- `test_update_changelog_head_dry_run_leaves_file_unchanged`
- `test_generate_head_changelog_updates_unreleased` (integration, wiremock)

Validation run locally:

- `cargo fmt --check` ✅
- `cargo clippy --all-targets --all-features -- -D warnings` ✅
- `cargo test` ✅ — 146 unit + 3 integration tests pass

Existing changelog and tagged-release tests continue to pass unchanged.

## Out of scope

- Literal `unreleased` CLI argument.
- Changes to `src/git.rs` range behavior — the existing `previous_tag(... "HEAD")` fallback to the most recent tag and `resolve_ref("HEAD")` already do the right thing for this feature.
- Any unrelated prompt/changelog refactors.

---

<details>
<summary>📋 Implementation Plan</summary>

# Plan: Support `communique generate HEAD --changelog` as an intentional `[Unreleased]` update

## Goal

Make the existing `HEAD` route an intentional way to generate unreleased changelog content:

```sh
communique generate HEAD --changelog
```

Expected behavior:

- Keep using existing git behavior to collect changes from the latest tag / explicit previous tag to `HEAD`.
- Treat `HEAD` as an unreleased target for display, prompt, and changelog behavior.
- Update the existing `## [Unreleased]` / `## Unreleased` section instead of inserting a literal `## [HEAD]` release entry.
- Do **not** add support for the literal `unreleased` argument in this change.

## Current evidence

<details>
<summary>Verified behavior from repo investigation</summary>

- `src/cli.rs` accepts `generate <tag>` as a required string argument; there is no special `unreleased` mode.
- `src/git.rs::previous_tag(repo_root, "HEAD")` does not find a tag named `HEAD`, then falls back to the most recent tag. This already gives the desired base tag when no previous tag override is supplied.
- `src/git.rs::resolve_ref(repo_root, "HEAD")` resolves `HEAD` as a valid git ref, giving the desired comparison end ref.
- `src/generate.rs::gather_context` stores the literal tag string in `Context.tag`.
- `src/prompt.rs::user_prompt` currently phrases the request as generating release notes for the literal tag, so `HEAD` appears as the release label.
- `src/generate.rs::update_changelog` currently uses `Context.tag` as the version identifier when inserting changelog content, so `HEAD --changelog` would create a literal `## [HEAD]`-style entry rather than updating `[Unreleased]`.
- `src/generate.rs` already has changelog tests covering insertion, new-file creation, dry-run behavior, and tail preservation.
- `tests/integration.rs` uses `wiremock`; unit tests use `MockLlmClient` / `TempRepo`, so the change can be validated without real API keys.

</details>

### In scope

1. Add intentional `HEAD` target handling for generated labels and changelog updates.
2. Preserve the existing git range mechanics.
3. Add focused tests for `HEAD --changelog`.
4. Add a clear guard for unsupported `HEAD --github-release` usage.
5. Keep changes surgical and preserve normal tagged release behavior.

### Out of scope

- Supporting `communique generate unreleased`.
- Reworking release-note generation architecture.
- Changing normal `communique generate vX.Y.Z --changelog` behavior.
- Publishing GitHub releases for `HEAD`.

### Phase 1 — Introduce a small target-mode helper

Add the smallest explicit abstraction needed to avoid leaking the literal `HEAD` into user-facing release labels.

Preferred minimal shape in `src/generate.rs`:

```rust
impl Context {
    fn is_unreleased_head(&self) -> bool {
        self.tag == "HEAD"
    }

    fn display_tag(&self) -> &str {
        if self.is_unreleased_head() {
            "Unreleased"
        } else {
            &self.tag
        }
    }
}
```

Rationale:

- Avoids adding a stored boolean that can drift out of sync with `Context.tag`.
- Keeps git/ref behavior unchanged: code that needs the actual ref should keep using `ctx.tag`.
- User-facing output, prompt text, and changelog mode can use `ctx.display_tag()` / `ctx.is_unreleased_head()`.

Quality gate after this phase:

- Compile the touched module/tests if possible: `cargo test --lib generate::` or the closest available focused test command.

### Phase 2 — Keep git range behavior unchanged

Do **not** change `src/git.rs` unless a focused test reveals a regression.

Implementation expectation:

- `communique generate HEAD` with no explicit previous tag continues to resolve the base as the latest tag via existing `previous_tag` fallback.
- `HEAD` continues to resolve as the end ref via existing `resolve_ref`.
- If the CLI supports an explicit previous tag override, preserve it for `HEAD`, e.g. `communique generate HEAD v1.0.3` should use `v1.0.3..HEAD`.

Add/adjust tests only if there is no direct coverage for the assumptions above:

- `previous_tag(repo_root, "HEAD")` returns the newest version tag in a tagged fixture.
- `resolve_ref(repo_root, "HEAD")` returns the current commit SHA / valid ref.

Quality gate after this phase:

```sh
cargo test previous_tag resolve_ref
```

If test names differ, run the relevant `src/git.rs` test subset or `cargo test --lib git`.

### Phase 3 — Update prompt and display labels for `HEAD`

In `src/prompt.rs`:

1. Add a boolean field to `UserPromptContext`, for example:

   ```rust
   pub is_unreleased_head: bool
   ```

2. When true, phrase the request as unreleased work since the previous tag, not as a release named `HEAD`.

   Example intent, not exact final copy:

   ```text
   Generate release notes for unreleased changes since v1.0.4.
   ```

3. Ensure the prompt does not contain `**HEAD**` as the release version in unreleased-head mode.

In `src/generate.rs` output/title handling:

- Use `ctx.display_tag()` anywhere the user-facing label is assembled.
- Keep using `ctx.tag` anywhere a git ref is required.
- Skip or adjust existing title normalization that forces titles into `{tag}: ...` format so it does not produce `HEAD: ...`.

Quality gate after this phase:

- Add prompt unit tests:
  - `HEAD` mode says “unreleased” and includes the previous tag.
  - `HEAD` mode does not label the release as `**HEAD**`.
  - Normal tag mode still says `**vX.Y.Z**`.

### Phase 4 — Branch changelog updates for `HEAD --changelog`

In `src/generate.rs::update_changelog`, branch early when `ctx.is_unreleased_head()` is true.

Normal tag behavior remains unchanged:

```text
## [Unreleased]

## [v1.0.4] - YYYY-MM-DD
...
```

New `HEAD` behavior:

```text
## [Unreleased]
<replace/update this section with generated changelog content>

## [v1.0.4] - YYYY-MM-DD
...
```

Recommended helper:

```rust
fn replace_unreleased_section(existing: &str, generated: &str) -> miette::Result<String>
```

Helper requirements:

- Work line-by-line rather than with fragile substring offsets.
- Accept both `## [Unreleased]` and `## Unreleased` headers, matching existing project parsing behavior.
- Preserve the exact unreleased header text found in the file.
- Preserve the file preamble and all released sections after the unreleased section.
- Replace only the body between the unreleased header and the next top-level `## ` heading.
- Normalize spacing to a predictable, readable form:

  ```md
  ## [Unreleased]

  ### Added
  - New item

  ## [v1.0.4] - 2026-04-01
  ```

- If `CHANGELOG.md` does not exist, keep the existing default creation behavior but populate the new `[Unreleased]` section.
- If an existing `CHANGELOG.md` lacks any Unreleased header, fail with a clear diagnostic instead of guessing where to insert it. This avoids silently corrupting nonstandard changelog layouts.

Defensive programming notes:

- Add `debug_assert!(ctx.is_unreleased_head())` inside the `HEAD`-only branch/helper caller where appropriate.
- Assert in tests that no `## [HEAD]`, `## HEAD`, or `HEAD:` changelog heading is introduced.
- Keep error handling user-facing for malformed/missing changelog sections; reserve assertions for impossible internal states.

Quality gate after this phase:

- Focused changelog tests pass.
- Existing normal-version changelog tests still pass unchanged.

### Phase 5 — Handle existing `[Unreleased]` context safely

Check `src/generate.rs::read_changelog_entry` / call sites used before LLM generation.

If the current code tries to read a changelog entry by `ctx.tag`, update the caller so that `HEAD` mode reads the existing `[Unreleased]` section instead of looking for `[HEAD]`.

Prompt behavior should make the intent clear:

- Existing `[Unreleased]` content is current draft material to reconcile/improve.
- The LLM should return the complete desired `[Unreleased]` section body, not a nested `## [Unreleased]` heading.

Acceptance test options:

- Unit test the section reader/helper if practical.
- Prompt test verifies existing unreleased content is described as unreleased draft content, not as an existing release entry for `HEAD`.

### Phase 6 — Reject unsupported GitHub release publishing for `HEAD`

Add an early guard for:

```sh
communique generate HEAD --github-release
```

Recommended behavior:

- Return a clear error before trying to fetch/update a GitHub release tagged `HEAD`.
- Suggested diagnostic:

  ```text
  --github-release cannot be used with HEAD because HEAD is an unreleased changelog target. Use --changelog, or generate notes for a real tag.
  ```

Reasoning:

- `HEAD` is not a release tag.
- Current behavior would search for a GitHub release tagged `HEAD`, which is misleading.
- This keeps the new `HEAD` mode scoped to unreleased notes/changelog content.

Quality gate after this phase:

- Add a CLI/options-level or generation-level test for the guard if the project has an easy pattern for this.
- Ensure normal `vX.Y.Z --github-release` behavior remains untouched.

### Unit tests in `src/generate.rs`

Add focused tests around the deterministic changelog helper and `update_changelog` path:

1. `test_replace_unreleased_section_bracketed_header`
   - Input has `## [Unreleased]` followed by a released version.
   - Output replaces only the unreleased body.
   - Released version content remains unchanged.
   - No `HEAD` heading is present.

2. `test_replace_unreleased_section_plain_header`
   - Same as above, but with `## Unreleased`.

3. `test_update_changelog_head_updates_unreleased_section`
   - Build `Context { tag: "HEAD", ... }`.
   - Use `ParsedOutput.changelog` with generated section body.
   - Verify `CHANGELOG.md` updates `[Unreleased]` in place.

4. `test_update_changelog_head_new_file_creates_unreleased_section`
   - No existing `CHANGELOG.md`.
   - Verify resulting file has `# Changelog`, `## [Unreleased]`, generated body, and no `## [HEAD]`.

5. `test_update_changelog_head_missing_unreleased_errors`
   - Existing changelog has released entries but no `Unreleased` heading.
   - Verify actionable error.

6. Existing normal changelog tests continue to pass:
   - Insert new version after `[Unreleased]` for real tags.
   - Dry-run does not write.
   - Tail preservation still works.

### Unit tests in `src/prompt.rs`

1. `test_user_prompt_unreleased_head_mode`
   - `tag = "HEAD"`, `prev_tag = "v1.0.4"`, `is_unreleased_head = true`.
   - Prompt contains “unreleased” and `v1.0.4`.
   - Prompt does not contain `**HEAD**`.

2. `test_user_prompt_tagged_release_mode_still_uses_tag`
   - `tag = "v1.0.5"`, `is_unreleased_head = false`.
   - Prompt still contains `**v1.0.5**` and previous release text.

### Integration test in `tests/integration.rs` if practical

Add or extend a wiremock-backed test that runs the CLI with:

```sh
communique generate HEAD --changelog --repo test/repo --provider openai --base-url <mock>
```

Verify:

- The command succeeds.
- `CHANGELOG.md` has updated `[Unreleased]` content.
- No literal `HEAD` changelog heading is created.
- Latest tagged release section remains below `[Unreleased]`.

If an integration test is too expensive for the first patch, keep it as a follow-up and rely on the unit tests plus existing integration coverage.

## Acceptance criteria

- `communique generate HEAD` still uses the latest tag / explicit previous tag as the lower bound and `HEAD` as the upper bound.
- `communique generate HEAD --changelog` updates `## [Unreleased]` / `## Unreleased` in place.
- `communique generate HEAD --changelog` never inserts a literal `## [HEAD]` / `## HEAD` changelog section.
- User-facing generated labels avoid `HEAD` and use “Unreleased” / “unreleased changes” wording.
- Normal tagged release generation is unchanged.
- `communique generate HEAD --github-release` fails with a clear message.
- New behavior is covered by unit tests, and normal changelog tests continue to pass.
- `cargo fmt --check`, `cargo clippy --all-targets --all-features -- -D warnings`, and `cargo test` pass.

### Automated quality gates between implementation phases

After changing prompt/display behavior:

```sh
cargo test --lib prompt
```

After changing changelog helpers:

```sh
cargo test --lib update_changelog
cargo test --lib replace_unreleased
```

If exact test filters differ, use the closest matching filters or run:

```sh
cargo test --lib
```

Before final handoff:

```sh
cargo fmt --check
cargo clippy --all-targets --all-features -- -D warnings
cargo test
```

If the repository expects the project wrapper, also run:

```sh
mise run lint
```

### Manual dogfooding recipe

Use a temporary git repository so no real project history or GitHub release is touched.

1. Build the local binary:

   ```sh
   cargo build
   ```

2. Create a temporary repo:

   ```sh
   TMP_DIR=$(mktemp -d)
   cd "$TMP_DIR"
   git init
   git config user.email "test@example.com"
   git config user.name "Test User"
   cat > CHANGELOG.md <<'EOF'
   # Changelog

   ## [Unreleased]

   ## [v1.0.4] - 2026-04-01
   ### Fixed
   - Previous released fix.
   EOF
   git add CHANGELOG.md
   git commit -m "chore: seed changelog"
   git tag v1.0.4
   echo "new behavior" > feature.txt
   git add feature.txt
   git commit -m "feat: add unreleased behavior"
   ```

3. Run the CLI against a mocked OpenAI-compatible endpoint using the existing `wiremock` integration-test pattern, or run against a real provider only if API keys are intentionally available.

   Example shape:

   ```sh
   /path/to/communique generate HEAD \
     --changelog \
     --repo test/repo \
     --provider openai \
     --model test-model \
     --base-url http://127.0.0.1:<mock-port> \
     --output notes.md
   ```

4. Verify the result:

   ```sh
   grep -n "Unreleased\|HEAD\|v1.0.4" CHANGELOG.md
   git diff -- CHANGELOG.md notes.md
   ```

   Expected:

   - `## [Unreleased]` contains the generated unreleased notes.
   - `## [v1.0.4]` remains below it.
   - No `## [HEAD]` / `## HEAD` section appears.
   - `notes.md` describes unreleased changes rather than a release named `HEAD`.

5. Negative dogfood:

   ```sh
   /path/to/communique generate HEAD --github-release --repo test/repo
   ```

   Expected: clear error explaining that `--github-release` cannot be used with `HEAD`.

### Evidence capture for reviewer

Because this is a CLI change, capture terminal evidence rather than browser screenshots:

- Record the manual dogfood session with `asciinema rec` if available, or with `script` as a fallback.
- Capture screenshots of:
  1. The temp repo setup and tag state (`git log --oneline --decorate --graph`).
  2. The successful `generate HEAD --changelog` run.
  3. The resulting `git diff -- CHANGELOG.md notes.md` showing `[Unreleased]` updated and no `[HEAD]` entry.
  4. The `HEAD --github-release` error.
- Attach the screenshots/video recording to the final implementation handoff if the environment supports attachments.

## Risks and mitigations

| Risk | Mitigation |
| --- | --- |
| Existing `[Unreleased]` content could be overwritten unexpectedly. | Make prompt include existing section content when available; document/test that the generated body replaces the section body. This mirrors “regenerate this section” semantics. |
| Changelog without an Unreleased section could be rearranged incorrectly. | Fail with a clear diagnostic for existing files missing the section; only auto-create when the file itself does not exist. |
| `HEAD` leaks into output titles or prompts. | Centralize label choice via `display_tag()` and add prompt/output tests that reject `HEAD` labels in unreleased mode. |
| Normal tag releases regress. | Keep branch condition narrowly scoped to `ctx.tag == "HEAD"`; rerun existing changelog and integration tests. |
| `--github-release` with `HEAD` behaves misleadingly. | Add an early explicit error. |

## Handoff notes for implementation

- Keep the diff small and avoid refactoring unrelated prompt/changelog code.
- Prefer deterministic changelog section replacement over a second LLM merge step for the `[Unreleased]` file edit.
- Match existing formatting style and error types in `src/generate.rs`.
- Use assertions only for internal invariants; use user-facing `miette` errors for malformed/missing changelog input.

</details>

---
_Generated with [`mux`](https://github.com/coder/mux) • Model: `anthropic:claude-opus-4-7` • Thinking: `max`_

### CODE REVIEW

File: src/generate.rs [hunk 1]
```diff
@@ -39,6 +39,67 @@ struct Context {
...
+            if raw_tag == display_tag {
+                None
+            } else {
+                release_title_description(&title, raw_tag)
+            }
+        })
+        .unwrap_or(title.as_str());
+
+    format!("{display_tag}: {description}")
+}
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/generate.rs:101]: Fixed in da2d7ea. The early-exit guard in `normalize_release_title` now catches `title.trim() == display_tag` as well, so a bare `"Unreleased"` from the LLM normalizes to `"Unreleased: changes"` instead of `"Unreleased: Unreleased"`. Added a regression test (`test_normalize_release_title_bare_display_tag`) plus an assertion that bare-tag normalization for tagged releases is unchanged.

File: src/generate.rs [hunk 2]
```diff
@@ -102,6 +149,8 @@ pub async fn run(opts: GenerateOptions) -> miette::Result<()> {
 }
 
 async fn gather_context(opts: &GenerateOptions, job: &Arc<ProgressJob>) -> miette::Result<Context> {
+    validate_generate_options(opts)?;
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/generate.rs]: Fixed in da2d7ea. Moved `validate_generate_options(&opts)?` to the very top of `run()`, before `ProgressJobBuilder::new()...start()`, and removed the redundant call from `gather_context`. `HEAD --github-release` now errors instantly without flashing a spinner. Added a smoke test that `run()` rejects this combination before any context/config loading.

File: src/generate.rs [hunk 3]
```diff
@@ -1388,6 +1673,126 @@ fn today_iso() -> String {
...
+        return Ok(None);
+    };
+
+    let body_start = lines[index].1;
+    let mut body_end = contents.len();
+    for (start, _, line) in lines.iter().skip(index + 1) {
+        if line.trim().starts_with("## ") {
+            body_end = *start;
+            break;
+        }
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/generate.rs:1809]: Good catch — fixed in da2d7ea. The system prompt does indeed instruct the model to use level-2 category headings (`## Added`, `## Fixed`, etc.), so a previously-generated `CHANGELOG.md` will have those inside `[Unreleased]`. Introduced an `is_version_header` helper that only treats version-style `## ` lines as the section terminator (lines starting with `[`, a digit, or `v`+digit). Category headings now stay inside the Unreleased body and are correctly replaced. Added regression tests for `## Added`/`## Fixed` inside the section and for `## v1.0.0` (no brackets) as the next version.

File: src/generate.rs [hunk 4]
```diff
@@ -1388,6 +1673,126 @@ fn today_iso() -> String {
...
+
+fn is_unreleased_header(line: &str) -> bool {
+    matches!(line.trim(), "## [Unreleased]" | "## Unreleased")
+}
+
+fn is_forbidden_unreleased_target_header(line: &str) -> bool {
+    matches!(
+        line.trim(),
+        "## [Unreleased]" | "## Unreleased" | "## [HEAD]" | "## HEAD"
+    )
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/generate.rs]: Fixed in da2d7ea. Broadened both `is_unreleased_header` and `is_forbidden_unreleased_target_header` to match `split_changelog`-style permissiveness, including linked variants like `## [Unreleased](https://example.com/compare/v1.0.0...HEAD)` and the corresponding `## [HEAD](...)` form. Added regression tests that linked Unreleased headers are detected and preserved exactly as found, and that linked HEAD headers are stripped from generated bodies.

File: src/generate.rs [hunk 5]
```diff
@@ -1388,6 +1673,126 @@ fn today_iso() -> String {
...
+    }
+
+    let tail = existing[section.body_end..].trim_start_matches(['\r', '\n']);
+    if !tail.is_empty() {
+        updated.push('\n');
+        updated.push_str(tail);
+    }
+
+    Ok(format!("{}\n", updated.trim_end()))
+}
```
[the remark this answers is not in the document]
Comment (ThomasK33) [reply src/generate.rs:1859]: Applied in da2d7ea. Adopted the suggested `trim_end()` + explicit `\n\n` separator pattern in `replace_unreleased_section`. This normalizes any CRLF prefix to LF and avoids the doubled-blank-lines case when the body is empty but a tail exists.

**Comment by ThomasK33** [comment]
Pushed `da2d7ea` addressing the review feedback. All comments resolved:

| # | Source | Issue | Fix |
| --- | --- | --- | --- |
| 1 | Greptile P2 | `normalize_release_title` produced `"Unreleased: Unreleased"` when the LLM emitted a bare `"Unreleased"` title | Extended early-exit guard to also match `title.trim() == display_tag` → `"Unreleased: changes"` |
| 2 | Greptile P2 | `validate_generate_options` ran inside `gather_context`, after the spinner started | Moved to top of `run()`; removed redundant call from `gather_context` |
| 3 | Copilot | `find_unreleased_section` ended at any `## ` heading, breaking on `## Added`/`## Fixed` (which the prompt encourages) | New `is_version_header` helper terminates only on `## [`, `## v<digit>`, or `## <digit>` |
| 4 | Copilot | `is_unreleased_header` only matched the two exact strings, missing linked variants like `## [Unreleased](url)` | Broadened both unreleased and forbidden-target matchers to align with `split_changelog`'s permissiveness |
| 5 | Gemini | `replace_unreleased_section` could produce mixed CRLF/LF endings or doubled blank lines | Adopted the suggested `trim_end()` + explicit `\n\n` separator pattern |

### Validation

- `cargo fmt --check` ✅
- `cargo clippy --all-targets --all-features -- -D warnings` ✅
- `cargo test` ✅ — **152 unit + 3 integration tests** (+6 unit regression tests)

Existing tagged-release behavior is unchanged; new tests target the exact edge cases each reviewer flagged.

---
_Generated with [`mux`](https://github.com/coder/mux) • Model: `anthropic:claude-opus-4-7` • Thinking: `max`_
